### DISCUSSION TITLE
docker CORS error

### DISCUSSION THREAD

**Discussion by hyh1618** [thomaxxl/safrs#134] [General]
Hi Thomaxxl,

Please help. I have code working fine, but when go into a docker instance, as the hostname is not normal: it has internal hostname not available to outside, and outside hostname not reachable by running process. I can bring up the swagger page, but any action will throw "CORS, Network Failure, URL scheme must be http or https for CORS request". I think the problem is that SAFRS is getting generated swagger files as local file, not through http. Any idea on a quick fix?

I did add line of "cors = CORS(app, resource ...), not help.

**Comment by thomaxxl** [comment]
Did you specify the `swagger_host` argument? Like in this example: https://github.com/thomaxxl/safrs/blob/cf75e8280d5d4870166cfbf89fe6399b1990bd1a/examples/demo_pythonanywhere_com.py#L276

**Reply by hyh1618** [reply]
Yes, I did refer to this sample. I have used internal hostname, 127.0.0.1, etc.  If you mean to use external hostname, I tried but it will fail to bind on it, as process inside docker can't see that hostname, even I can connect to this instance by that hostname.

**Reply by hyh1618** [reply]
what resources it will open as File? can we make them through http? I do have another web server in same instance.

**Reply by thomaxxl** [reply]
The `SAFRSAPI` `swagger_host` argument should be different from the `host` argument passed to `app.run()` in your case. Can you verify this please?

**Reply by hyh1618** [reply]
I did, used "0.0.0.0" and None inside SAFRSAPI(), and run with other hostname port, still same error:

CORS
Network Failure
URL scheme must be "http" or "https" for CORS request

**Comment by thomaxxl** [comment]
Some debug tips:

1) The CORS error may be misleading: if the IP is not reachable (for example when DNS doesn't resolve), the browser console also shows this error. Please verify the actual request from the network tab in the browser dev tools. 

2) `SAFRSAPI` `swagger_host` should reference the external ip or DNS name, the `app.run()` `host` argument refers to the local bind IP address (typically `0.0.0.0`)

3) You can check if the API endpoints are available, for example if you have a `/api/users` endpoint, then you should check if you get a response for http://x.x.x.x/api/users where x.x.x.x is the external IP of the container

4) You can check the swagger.yaml: paste it in https://editor.swagger.io/ , replace the IP or hostname by the external IP of the container, then check if it works (note this may lead to actual CORS errors, i.e. the `Access-Control-Allow-Origin` is not set or restricted)

**Reply by hyh1618** [reply]
I did use external and internal per #2.  Now its not CORS error, its "internal server error". Flask abort, safrs/jsonapi_formatting.py throw GenericError(f"{exc}")

**Reply by thomaxxl** [reply]
This means there is an error in your implementation. The stacktrace in the container logs contains more information.

**Reply by hyh1618** [reply]
there is no meaningful error msg, and same code works fine in my local. The error included end point url is using the external hostname. Is it because that external is not resolvable inside docker instance?

**Reply by thomaxxl** [reply]
The error you see comes from the location where the actual sqlalchemy query is executed. Maybe there is a problem with your db connection but I can't tell without more information.
What db backend do you use?

**Reply by hyh1618** [reply]
It's sqlite. I can start the app and see the home page, just when I run any end point it throw this error. The sqlite db did created, and my other program in same docker tested sqlite OK. The problem should be with safrs/sqlalchemy/flask

**Comment by L4db** [comment]
Maybe increase the debug level in `app.config`, or by setting the `FLASK_DEBUG` env var.

Thomas Pollet
________________________________

**Reply by hyh1618** [reply]
I did that.  not much besides: safrs.errors.GenericError, ERROR:safrs.safrs_init. ..., KeyError: <weakref at ..., to Flask ar ..

**Reply by hyh1618** [reply]
BTW, why your sample code use swagger_host="0.0.0.0"?

**Reply by thomaxxl** [reply]
Oh, you get that error. https://github.com/thomaxxl/safrs/issues/127

Can you downgrade Flask-SQLAlchemy to v2.5.1?

**Comment by hyh1618** [comment]
Thanks, it fixed that error!
